### PR TITLE
hide ctxtypes and use connect interceptor instead

### PR DESCRIPTION
**Pull request by doodlesbykumbi**

### CODE REVIEW

File: connect_interceptor.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+type ConnectInterceptor struct {
+	// PreLoginResponse is used to obtain PreLogin Response fields
+	PreLoginResponse chan map[uint8][]byte
+	// ClientLogin is used to pass the client Login
+	ClientLogin chan Login
+	// ServerLoginAck is used to pass the server LoginAck
+	ServerLoginAck chan LoginAckStruct
+	// ServerError is used to pass the server error
+	ServerError chan Error
+}
```
Comment (jonahx) [inline connect_interceptor.go]: I really like this abstraction. 

Not a blocker, but I would consider a simpler name like `ConnectionSteps` or `ProtocolSteps` or `ProtocolStepResponses` (etc).    Going back to the conceptual purity argument that convinced me to put this stuff here: If this is part of an extended public API for the driver, the fact that we're hooking into the `Connect` method is an impl detail we don't care about.  

It's really providing us a way to get the responses from the various protocol steps.

File: connect_interceptor.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+func (c *ConnectInterceptor) Close()  {
```
Comment (diverdane) [inline connect_interceptor.go]: golint nit: Exported Close() function needs a comment, and `gofmt` is complaining about the extra space before '{'.

File: connect_interceptor.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		// Create a channel for sending the client login to the driver through the context
+		ClientLogin: make(chan Login),
+		// Create a channel for receiving a server loginAck from the driver through the context
+		ServerLoginAck: make(chan LoginAckStruct),
+		// Create a channel for receiving a server error from the driver through the context
+		ServerError: make(chan Error),
+	}
+}
+
+// NewContext returns a new Context that carries value ci.
```
Comment (diverdane) [inline connect_interceptor.go]: s/NewContext/NewContextWithConnectInterceptor/

File: connect_interceptor.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+func NewContextWithConnectInterceptor(ctx context.Context, u *ConnectInterceptor) context.Context {
+	return context.WithValue(ctx, connectInterceptorKey, u)
+}
+
+// FromContext returns the ConnectInterceptor value stored in ctx, if any.
```
Comment (diverdane) [inline connect_interceptor.go]: s/FromContext/ConnectInterceptorFromContext/

File: connect_interceptor.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+func (c *ConnectInterceptor) Close()  {
+	close(c.PreLoginResponse)
+	close(c.ClientLogin)
+	close(c.ServerLoginAck)
+	close(c.ServerError)
+}
+
+func NewConnectInterceptor() *ConnectInterceptor {
```
Comment (diverdane) [inline connect_interceptor.go:22]: 'nother golint nit: Exported NewConnectInterceptor() should have comment.

File: tds.go [hunk 6]
```diff
@@ -204,11 +197,16 @@ func WritePreloginWithPacketType(w *TdsBuffer, fields map[uint8][]byte, packetTy
...
 
-func readPrelogin(r *TdsBuffer) (map[uint8][]byte, error) {
-	return ReadPreloginWithPacketType(r, PackReply)
+// http://msdn.microsoft.com/en-us/library/dd357559.aspx
+func writePrelogin(w *TdsBuffer, fields map[uint8][]byte) error {
+	return writePreloginWithPacketType(w, fields, PackPrelogin)
 }
 
-func ReadPreloginWithPacketType(r *TdsBuffer, expectedPacketType packetType) (map[uint8][]byte, error) {
+func WritePreloginResponse(w *TdsBuffer, fields map[uint8][]byte) error {
```
Comment (diverdane) [inline tds.go]: golint nits: exported WritePreloginResponse() and ReadPreloginRequest() need comments.

File: tds.go [hunk 7]
```diff
@@ -986,11 +992,8 @@ initiate_connection:
 	TypeFlags := p.typeFlags
 
 	// Replaces params with values from the client Login
-	if clientLoginChan := ctx.Value(ctxtypes.ClientLoginKey); clientLoginChan != nil {
-		// A panic will never occur here unless the calling code is wrong
-		// TODO: perhaps there should be a timeout to this
-		//  suppose the channel was forever empty == leak
-		clientLogin := <- clientLoginChan.(chan Login)
+	if connectInterceptor != nil && connectInterceptor.ClientLogin != nil {
+		clientLogin := <- connectInterceptor.ClientLogin
```
Comment (diverdane) [inline tds.go]: gofmt nit: should be:
```
clientLogin := <-connectInterceptor.ClientLogin
```

**Review by diverdane — COMMENTED** [review]
Fantastic stuff. Just a few gofmt/golint nits.

**Review by BradleyBoutcher — APPROVED** [review]

**Pull request merged**